application/common/components/Emailer.php

```diff
@@ -410,24 +421,66 @@ public function sendDelayedMfaRelatedEmails()
     }
 
     /**
-     *
      * Whether the user has already been sent this type of email in the last X days
      *
      * @param int $userId
      * @param string $messageType
      * @return bool
      */
-    public function hasReceivedMessageRecently($userId, string $messageType)
+    public function hasUserReceivedMessageRecently($userId, string $messageType): bool
     {
-        $latestEmail = EmailLog::find()->where(['user_id' => $userId, 'message_type' => $messageType])
-            ->orderBy('sent_utc DESC')->one();
+        $user = User::findOne($userId);
+        if (empty($user)) {
+            \Yii::warning(sprintf(
+                'User (%s) not found when checking for recent emails',
+                $userId
+            ));
+            return false;
+        }
+        return $this->hasAddressReceivedMessageRecently($user->email, $messageType);
+    }
+
+    public function hasAddressReceivedMessageRecently(?string $toAddress, string $messageType): bool
+    {
+        if (empty($toAddress)) {
+            \Yii::warning('Not given any email address when checking for recent emails');
```

Comment:
Not sure this is the best idea, but I'll suggest anyway. Remove this warning and then you can drop the duplicate check in `shouldSendAbandonedUsersMessage`.
```suggestion
```

Reply:
I'm not sure I understand. This seems like a check that this function should be responsible for doing, to ensure it properly handles any unacceptable values it might be passed, rather than depending on all of the other code that might call this to do the data validation for it. I'd also like to avoid too tightly coupling this function's logic with that of the `shouldSendAbandonedUsersMessage()` function.

Reply:
My comment was simply reducing complexity, with little to no consideration of coupling. Do what you think is best.